File: ios/loveaceios/Network/AUFEConnection.swift
import Foundation
import CommonCrypto
import os

private let logger = Logger(subsystem: "tech.loveace.loveaceios", category: "AUFEConnection")

actor AUFEConnection {
    let userId: String
    private let ecPassword: String
    private let password: String

    var client: HTTPClient!
    var simpleClient: HTTPClient!
    var noRedirectClient: HTTPClient!

    private(set) var twfId: String?
    private var ecLogged = false
    private var uaapLogged = false

    static let serverURL = "https://vpn2.aufe.edu.cn"
    static let uaapLoginURL =
        "http://uaap-aufe-edu-cn.vpn2.aufe.edu.cn:8118/cas/login?service=http%3A%2F%2Fjwcxk2.aufe.edu.cn%2Fj_spring_cas_security_check"
    static let timeout: TimeInterval = 60

    init(userId: String, ecPassword: String, password: String) {
        self.userId = userId
        self.ecPassword = ecPassword
        self.password = password
    }

    func startClient() {
        let storage = HTTPCookieStorage.shared
        storage.cookieAcceptPolicy = .always
        if let oldCookies = storage.cookies {
            for c in oldCookies { storage.deleteCookie(c) }
        }
        client = HTTPClient(baseUrl: Self.serverURL, timeoutInterval: Self.timeout, cookieStorage: storage)
        simpleClient = HTTPClient(baseUrl: Self.serverURL, timeoutInterval: Self.timeout, cookieStorage: storage)
        noRedirectClient = HTTPClient(baseUrl: Self.serverURL, timeoutInterval: Self.timeout,
                                      followRedirects: false, cookieStorage: storage)
    }

    func setOnSessionExpired(_ callback: @escaping @Sendable () -> Void) async {
        await client.setSessionExpired(callback)
        await simpleClient.setSessionExpired(callback)
    }

    // MARK: - Heartbeat

    func heartbeat() async -> Bool {
        do {
            let (data, _) = try await client.get("\(Self.serverURL)/por/login_auth.csp?apiversion=1")
            let body = String(data: data, encoding: .utf8) ?? ""
            let alive = body.contains("<TwfID>")
            logger.debug("💓 Heartbeat: \(alive ? "alive" : "expired")")
            return alive
        } catch {
            logger.warning("💓 Heartbeat failed: \(error.localizedDescription)")
            return false
        }
    }

    // MARK: - Reconnect

    func reconnect() async -> Bool {
        logger.info("🔄 Reconnecting...")
        startClient()
        let ec = await ecLogin()
        guard ec.success else { logger.error("🔄 EC re-login failed"); return false }
        let uaap = await uaapLogin()
        guard uaap.success else { logger.error("🔄 UAAP re-login failed"); return false }
        logger.info("✅ Reconnect succeeded")
        return true
    }

    // MARK: - EC Login

    func ecLogin() async -> ECLoginStatus {
        do {
            return try await performEcLogin()
        } catch {
            logger.error("EC login error: \(error.localizedDescription)")
            return ECLoginStatus(failUnknownError: true)
        }
    }

    private func performEcLogin() async throws -> ECLoginStatus {
        logger.info("🔑 EC Login Step 1: fetching auth params...")
        let (data, _) = try await client.get("\(Self.serverURL)/por/login_auth.csp?apiversion=1")
        let body = String(data: data, encoding: .utf8) ?? ""
        if body.isEmpty {
            logger.error("❌ EC auth response is empty")
            return ECLoginStatus(failNetworkError: true)
        }
        logger.info("📄 EC auth response: \(body.prefix(300))")

        guard let twfMatch = body.range(of: "(?<=<TwfID>).+?(?=</TwfID>)", options: .regularExpression) else {
            logger.error("❌ TwfID not found in response")
            return ECLoginStatus(failNotFoundTwfid: true)
        }
        twfId = String(body[twfMatch])
        logger.info("✅ TwfID: \(self.twfId ?? "nil")")

        guard let rsaKeyMatch = body.range(of: "(?<=<RSA_ENCRYPT_KEY>).+?(?=</RSA_ENCRYPT_KEY>)", options: .regularExpression) else {
            logger.error("❌ RSA key not found")
            return ECLoginStatus(failNotFoundRsaKey: true)
        }
        let rsaKey = String(body[rsaKeyMatch])
        logger.info("✅ RSA key length: \(rsaKey.count)")

        guard let rsaExpMatch = body.range(of: "(?<=<RSA_ENCRYPT_EXP>).+?(?=</RSA_ENCRYPT_EXP>)", options: .regularExpression) else {
            logger.error("❌ RSA exponent not found")
            return ECLoginStatus(failNotFoundRsaExp: true)
        }
        let rsaExp = String(body[rsaExpMatch])
        logger.info("✅ RSA exponent: \(rsaExp)")

        guard let csrfMatch = body.range(of: "(?<=<CSRF_RAND_CODE>).+?(?=</CSRF_RAND_CODE>)", options: .regularExpression) else {
            logger.error("❌ CSRF code not found")
            return ECLoginStatus(failNotFoundCsrfCode: true)
        }
        let csrfCode = String(body[csrfMatch])
        logger.info("✅ CSRF code: \(csrfCode)")

        let passwordToEncrypt = "\(ecPassword)_\(csrfCode)"
        let encryptedPassword = CryptoHelper.rsaEncrypt(plaintext: passwordToEncrypt, modulusHex: rsaKey, exponentStr: rsaExp)
        logger.info("🔐 Encrypted password length: \(encryptedPassword.count), empty: \(encryptedPassword.isEmpty)")

        logger.info("🔑 EC Login Step 2: posting credentials...")
        let (loginData, _) = try await client.post(
            "\(Self.serverURL)/por/login_psw.csp?anti_replay=1&encrypt=1&type=cs",
            formData: [
                "svpn_rand_code": "",
                "mitm": "",
                "svpn_req_randcode": csrfCode,
                "svpn_name": userId,
                "svpn_password": encryptedPassword
            ],
            headers: ["Cookie": "TWFID=\(twfId!)"]
        )
        let loginBody = String(data: loginData, encoding: .utf8) ?? ""
        logger.info("📄 EC login response: \(loginBody.prefix(500))")

        if loginBody.contains("<Result>1</Result>") {
            await client.setCookie(name: "TWFID", value: twfId!, domain: ".vpn2.aufe.edu.cn")
            ecLogged = true
            await simpleClient.copyCookies(from: client)
            await noRedirectClient.copyCookies(from: client)
            logger.info("✅ EC Login succeeded")
            return ECLoginStatus(success: true)
        } else if loginBody.contains("Invalid username or password!") {
            logger.error("❌ EC Login: invalid credentials")
            return ECLoginStatus(failInvalidCredentials: true)
        } else if loginBody.contains("[CDATA[maybe attacked]]") || loginBody.contains("CAPTCHA required") {
            logger.error("❌ EC Login: maybe attacked / captcha required")
            return ECLoginStatus(failMaybeAttacked: true)
        }
        logger.error("❌ EC Login: unknown error, response snippet: \(loginBody.prefix(200))")
        return ECLoginStatus(failUnknownError: true)
    }

    // MARK: - UAAP Login

    func uaapLogin() async -> UAAPLoginStatus {
        do {
            return try await performUaapLogin()
        } catch {
            logger.error("UAAP login error: \(error.localizedDescription)")
            return UAAPLoginStatus(failUnknownError: true)
        }
    }

    private func performUaapLogin() async throws -> UAAPLoginStatus {
        logger.info("🔑 UAAP Login Step 1: fetching CAS page...")
        let (data, _) = try await client.get(Self.uaapLoginURL)
        let body = String(data: data, encoding: .utf8) ?? ""
        if body.isEmpty {
            logger.error("❌ UAAP CAS page is empty")
            return UAAPLoginStatus(failNetworkError: true)
        }

        guard let ltMatch = body.range(of: #"(?<=name="lt" value=").+?(?=")"#, options: .regularExpression) else {
            logger.error("❌ lt not found. Body snippet: \(body.prefix(500))")
            return UAAPLoginStatus(failNotFoundLt: true)
        }
        let ltValue = String(body[ltMatch])
        logger.info("✅ lt value: \(ltValue.prefix(20))... (len=\(ltValue.count))")

        guard let execMatch = body.range(of: #"(?<=name="execution" value=").+?(?=")"#, options: .regularExpression) else {
            logger.error("❌ execution not found")
            return UAAPLoginStatus(failNotFoundExecution: true)
        }
        let executionValue = String(body[execMatch])
        logger.info("✅ execution: \(executionValue.prefix(20))...")

        let encryptedPassword = CryptoHelper.desEncrypt(plaintext: password, key: ltValue)
        logger.info("🔐 DES encrypted password: \(encryptedPassword.prefix(30))... (len=\(encryptedPassword.count))")

        logger.info("🔑 UAAP Login Step 2: posting credentials...")
        let (loginData, loginResponse) = try await client.post(
            Self.uaapLoginURL,
            formData: [
                "username": userId,
                "password": encryptedPassword,
                "lt": ltValue,
                "execution": executionValue,
                "_eventId": "submit",
                "submit": "LOGIN"
            ]
        )
        let loginBody = String(data: loginData, encoding: .utf8) ?? ""
        let responseUrl = loginResponse.url?.absoluteString ?? ""
        logger.info("📄 UAAP response URL: \(responseUrl)")
        logger.info("📄 UAAP response size: \(loginBody.count), contains '用户名或密码错误': \(loginBody.contains("用户名或密码错误")), contains 'ticket': \(responseUrl.contains("ticket="))")

        if loginBody.contains("Invalid username or password") || loginBody.contains("用户名或密码错误") {
            logger.error("❌ UAAP Login: invalid credentials")
            return UAAPLoginStatus(failInvalidCredentials: true)
        }
        if responseUrl.hasPrefix("http://jwcxk2") || responseUrl.contains("ticket=") {
            uaapLogged = true
            await simpleClient.copyCookies(from: client)
            await noRedirectClient.copyCookies(from: client)
            logger.info("✅ UAAP Login succeeded")
            return UAAPLoginStatus(success: true)
        }
        logger.error("❌ UAAP Login: unknown error. responseUrl=\(responseUrl), body snippet: \(loginBody.prefix(300))")
        return UAAPLoginStatus(failUnknownError: true)
    }

    var isHealthy: Bool { ecLogged && uaapLogged }
}

// MARK: - HTTPClient extension for session expired

extension HTTPClient {
    func setSessionExpired(_ callback: @escaping @Sendable () -> Void) {
        self.onSessionExpired = callback
    }
}

// MARK: - CryptoHelper

enum CryptoHelper {
    static func rsaEncrypt(plaintext: String, modulusHex: String, exponentStr: String) -> String {
        guard let plainData = plaintext.data(using: .utf8) else { return "" }
        guard let modulusData = hexToData(modulusHex) else { return "" }
        guard let exponent = UInt32(exponentStr) else { return "" }

        var expBytes = withUnsafeBytes(of: exponent.bigEndian) { Array($0) }
        while expBytes.first == 0 && expBytes.count > 1 { expBytes.removeFirst() }
        let exponentBytes = Data(expBytes)

        let keyData = buildDERPublicKey(modulus: modulusData, exponent: exponentBytes)

        let attributes: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeyClass as String: kSecAttrKeyClassPublic,
            kSecAttrKeySizeInBits as String: modulusData.count * 8,
        ]
        var error: Unmanaged<CFError>?
        guard let secKey = SecKeyCreateWithData(keyData as CFData, attributes as CFDictionary, &error) else {
            return ""
        }
        guard let encrypted = SecKeyCreateEncryptedData(secKey, .rsaEncryptionPKCS1, plainData as CFData, &error) else {
            return ""
        }
        return (encrypted as Data).map { String(format: "%02x", $0) }.joined()
    }

    private static func hexToData(_ hex: String) -> Data? {
        var data = Data()
        var idx = hex.startIndex
        while idx < hex.endIndex {
            let next = hex.index(idx, offsetBy: 2, limitedBy: hex.endIndex) ?? hex.endIndex
            guard next != idx else { break }
            guard let byte = UInt8(hex[idx..<next], radix: 16) else { return nil }
            data.append(byte)
            idx = next
        }
        return data
    }

    private static func buildDERPublicKey(modulus: Data, exponent: Data) -> Data {
        var modulusBytes = modulus
        if modulusBytes.first! >= 0x80 { modulusBytes.insert(0x00, at: 0) }
        var exponentBytes = exponent
        if exponentBytes.first! >= 0x80 { exponentBytes.insert(0x00, at: 0) }

        let modTLV = derInteger(modulusBytes)
        let expTLV = derInteger(exponentBytes)
        let sequence = derSequence(modTLV + expTLV)

        let bitString = Data([0x03]) + derLength(sequence.count + 1) + Data([0x00]) + sequence
        let algorithmOID: [UInt8] = [
            0x30, 0x0D, 0x06, 0x09, 0x2A, 0x86, 0x48, 0x86,
            0xF7, 0x0D, 0x01, 0x01, 0x01, 0x05, 0x00
        ]
        let outerSequence = derSequence(Data(algorithmOID) + bitString)
        return outerSequence
    }

    private static func derInteger(_ data: Data) -> Data {
        Data([0x02]) + derLength(data.count) + data
    }

    private static func derSequence(_ data: Data) -> Data {
        Data([0x30]) + derLength(data.count) + data
    }

    private static func derLength(_ length: Int) -> Data {
        if length < 0x80 { return Data([UInt8(length)]) }
        if length < 0x100 { return Data([0x81, UInt8(length)]) }
        return Data([0x82, UInt8(length >> 8), UInt8(length & 0xFF)])
    }

    static func desEncrypt(plaintext: String, key: String) -> String {
        guard let plainData = plaintext.data(using: .utf8) else { return "" }
        var keyBytes = Array(key.utf8)
        if keyBytes.count > 8 { keyBytes = Array(keyBytes.prefix(8)) }
        while keyBytes.count < 8 { keyBytes.append(0) }

        var tripleKey = [UInt8](repeating: 0, count: 24)
        for i in 0..<8 { tripleKey[i] = keyBytes[i]; tripleKey[i+8] = keyBytes[i]; tripleKey[i+16] = keyBytes[i] }

        let bufferSize = plainData.count + kCCBlockSize3DES
        var buffer = [UInt8](repeating: 0, count: bufferSize)
        var numBytesEncrypted: size_t = 0

        let status = plainData.withUnsafeBytes { plainBytes in
            CCCrypt(CCOperation(kCCEncrypt), CCAlgorithm(kCCAlgorithm3DES),
                    CCOptions(kCCOptionPKCS7Padding | kCCOptionECBMode),
                    tripleKey, kCCKeySize3DES, nil,
                    plainBytes.baseAddress, plainData.count,
                    &buffer, bufferSize, &numBytesEncrypted)
        }

        guard status == kCCSuccess else { return "" }
        return Data(buffer.prefix(numBytesEncrypted)).base64EncodedString()
    }
}

File: android/app/src/main/java/tech/loveace/appv3/data/network/AUFEConnection.kt
package tech.loveace.appv3.data.network

import android.util.Base64
import android.util.Log
import tech.loveace.appv3.data.model.ECLoginStatus
import tech.loveace.appv3.data.model.UAAPLoginStatus
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.bouncycastle.crypto.encodings.PKCS1Encoding
import org.bouncycastle.crypto.engines.DESedeEngine
import org.bouncycastle.crypto.engines.RSAEngine
import org.bouncycastle.crypto.paddings.PKCS7Padding
import org.bouncycastle.crypto.paddings.PaddedBufferedBlockCipher
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.crypto.params.RSAKeyParameters
import java.math.BigInteger

/**
 * AUFE 教务系统连接器
 * 处理 EC(VPN) 和 UAAP(CAS) 双重认证
 */
class AUFEConnection(
    val userId: String,
    private val ecPassword: String,
    private val password: String,
) {
    lateinit var client: HttpClient
        private set
    lateinit var simpleClient: HttpClient
        private set
    lateinit var noRedirectClient: HttpClient
        private set

    var twfId: String? = null
        private set
    private var ecLogged = false
    private var uaapLogged = false

    fun startClient() {
        client = HttpClient(baseUrl = SERVER_URL, timeoutMs = TIMEOUT)
        simpleClient = HttpClient(baseUrl = SERVER_URL, timeoutMs = TIMEOUT)
        noRedirectClient = HttpClient(baseUrl = SERVER_URL, timeoutMs = TIMEOUT, followRedirects = false)
    }

    /** 设置 session 过期回调（所有 client 共享） */
    fun setOnSessionExpired(callback: () -> Unit) {
        client.onSessionExpired = callback
        simpleClient.onSessionExpired = callback
        // noRedirectClient 不设置，因为它本身就处理重定向
    }

    /** 心跳：轻量 GET 保持 VPN session 活跃 */
    fun heartbeat(): Boolean {
        return try {
            val response = client.get("$SERVER_URL/por/login_auth.csp?apiversion=1")
            val body = response.body?.string() ?: ""
            response.close()
            val alive = body.contains("<TwfID>")
            Log.d(TAG, "💓 Heartbeat: ${if (alive) "alive" else "expired"}")
            alive
        } catch (e: Exception) {
            Log.w(TAG, "💓 Heartbeat failed: ${e.message}")
            false
        }
    }

    /** 完整重连：EC + UAAP */
    suspend fun reconnect(): Boolean = withContext(Dispatchers.IO) {
        try {
            Log.i(TAG, "🔄 Reconnecting...")
            close()
            startClient()
            val ec = ecLogin()
            if (!ec.success) { Log.e(TAG, "🔄 EC re-login failed"); return@withContext false }
            val uaap = uaapLogin()
            if (!uaap.success) { Log.e(TAG, "🔄 UAAP re-login failed"); return@withContext false }
            Log.i(TAG, "✅ Reconnect succeeded")
            true
        } catch (e: Exception) {
            Log.e(TAG, "🔄 Reconnect error", e)
            false
        }
    }

    suspend fun ecLogin(): ECLoginStatus = withContext(Dispatchers.IO) {
        try {
            performEcLogin()
        } catch (e: Exception) {
            Log.e(TAG, "EC login error", e)
            ECLoginStatus(failUnknownError = true)
        }
    }

    private fun performEcLogin(): ECLoginStatus {
        // 1. Get auth params
        val response = client.get("$SERVER_URL/por/login_auth.csp?apiversion=1")
        val body = response.body?.string() ?: return ECLoginStatus(failNetworkError = true)

        // 2. Extract TwfID
        val twfMatch = Regex("<TwfID>(.*?)</TwfID>").find(body)
            ?: return ECLoginStatus(failNotFoundTwfid = true)
        twfId = twfMatch.groupValues[1]

        // 3. Extract RSA key
        val rsaKeyMatch = Regex("<RSA_ENCRYPT_KEY>(.*?)</RSA_ENCRYPT_KEY>").find(body)
            ?: return ECLoginStatus(failNotFoundRsaKey = true)
        val rsaKey = rsaKeyMatch.groupValues[1]

        // 4. Extract RSA exponent
        val rsaExpMatch = Regex("<RSA_ENCRYPT_EXP>(.*?)</RSA_ENCRYPT_EXP>").find(body)
            ?: return ECLoginStatus(failNotFoundRsaExp = true)
        val rsaExp = rsaExpMatch.groupValues[1]

        // 5. Extract CSRF code
        val csrfMatch = Regex("<CSRF_RAND_CODE>(.*?)</CSRF_RAND_CODE>").find(body)
            ?: return ECLoginStatus(failNotFoundCsrfCode = true)
        val csrfCode = csrfMatch.groupValues[1]

        // 6. RSA encrypt password
        val passwordToEncrypt = "${ecPassword}_$csrfCode"
        val encryptedPassword = rsaEncrypt(passwordToEncrypt, rsaKey, rsaExp)

        // 7. Login
        val loginResponse = client.post(
            "$SERVER_URL/por/login_psw.csp?anti_replay=1&encrypt=1&type=cs",
            formData = mapOf(
                "svpn_rand_code" to "",
                "mitm" to "",
                "svpn_req_randcode" to csrfCode,
                "svpn_name" to userId,
                "svpn_password" to encryptedPassword,
            ),
            headers = mapOf("Cookie" to "TWFID=$twfId"),
        )
        val loginBody = loginResponse.body?.string() ?: return ECLoginStatus(failNetworkError = true)

        return when {
            loginBody.contains("<Result>1</Result>") -> {
                client.cookieJar.setCookie("TWFID", twfId!!, "vpn2.aufe.edu.cn")
                ecLogged = true
                simpleClient.copyCookiesFrom(client)
                noRedirectClient.copyCookiesFrom(client)
                ECLoginStatus(success = true)
            }
            loginBody.contains("Invalid username or password!") ->
                ECLoginStatus(failInvalidCredentials = true)
            loginBody.contains("[CDATA[maybe attacked]]") || loginBody.contains("CAPTCHA required") ->
                ECLoginStatus(failMaybeAttacked = true)
            else -> ECLoginStatus(failUnknownError = true)
        }
    }

    suspend fun uaapLogin(): UAAPLoginStatus = withContext(Dispatchers.IO) {
        try {
            performUaapLogin()
        } catch (e: Exception) {
            Log.e(TAG, "UAAP login error", e)
            UAAPLoginStatus(failUnknownError = true)
        }
    }

    private fun performUaapLogin(): UAAPLoginStatus {
        // 1. Get login page
        val response = client.get(UAAP_LOGIN_URL)
        val body = response.body?.string() ?: return UAAPLoginStatus(failNetworkError = true)

        // 2. Extract lt
        val ltMatch = Regex("""name="lt" value="(.*?)"""").find(body)
            ?: return UAAPLoginStatus(failNotFoundLt = true)
        val ltValue = ltMatch.groupValues[1]

        // 3. Extract execution
        val execMatch = Regex("""name="execution" value="(.*?)"""").find(body)
            ?: return UAAPLoginStatus(failNotFoundExecution = true)
        val executionValue = execMatch.groupValues[1]

        // 4. DES encrypt password
        val encryptedPassword = desEncrypt(password, ltValue)

        // 5. Submit login form
        val loginResponse = client.post(
            UAAP_LOGIN_URL,
            formData = mapOf(
                "username" to userId,
                "password" to encryptedPassword,
                "lt" to ltValue,
                "execution" to executionValue,
                "_eventId" to "submit",
                "submit" to "LOGIN",
            ),
        )
        val loginBody = loginResponse.body?.string() ?: return UAAPLoginStatus(failNetworkError = true)
        val responseUrl = loginResponse.request.url.toString()

        return when {
            loginBody.contains("Invalid username or password") ||
                loginBody.contains("用户名或密码错误") ->
                UAAPLoginStatus(failInvalidCredentials = true)
            responseUrl.startsWith("http://jwcxk2") || responseUrl.contains("ticket=") -> {
                uaapLogged = true
                simpleClient.copyCookiesFrom(client)
                noRedirectClient.copyCookiesFrom(client)
                UAAPLoginStatus(success = true)
            }
            else -> UAAPLoginStatus(failUnknownError = true)
        }
    }

    fun isHealthy(): Boolean = ecLogged && uaapLogged

    suspend fun close() = withContext(Dispatchers.IO) {
        runCatching { client.close() }
        runCatching { simpleClient.close() }
        runCatching { noRedirectClient.close() }
    }

    // ==================== Crypto ====================

    private fun rsaEncrypt(plaintext: String, modulusHex: String, exponentStr: String): String {
        val modulus = BigInteger(modulusHex, 16)
        val exponent = BigInteger(exponentStr)
        val publicKey = RSAKeyParameters(false, modulus, exponent)
        val encryptor = PKCS1Encoding(RSAEngine())
        encryptor.init(true, publicKey)
        val plainBytes = plaintext.toByteArray(Charsets.UTF_8)
        val encrypted = encryptor.processBlock(plainBytes, 0, plainBytes.size)
        return encrypted.joinToString("") { "%02x".format(it) }
    }

    private fun desEncrypt(plaintext: String, key: String): String {
        var keyBytes = key.toByteArray(Charsets.UTF_8)
        keyBytes = when {
            keyBytes.size > 8 -> keyBytes.copyOf(8)
            keyBytes.size < 8 -> keyBytes.copyOf(8)
            else -> keyBytes
        }
        // TripleDES: repeat 8-byte key 3 times
        val tripleKey = ByteArray(24)
        System.arraycopy(keyBytes, 0, tripleKey, 0, 8)
        System.arraycopy(keyBytes, 0, tripleKey, 8, 8)
        System.arraycopy(keyBytes, 0, tripleKey, 16, 8)

        val cipher = PaddedBufferedBlockCipher(DESedeEngine(), PKCS7Padding())
        cipher.init(true, KeyParameter(tripleKey))
        val input = plaintext.toByteArray(Charsets.UTF_8)
        val output = ByteArray(cipher.getOutputSize(input.size))
        val len = cipher.processBytes(input, 0, input.size, output, 0)
        val finalLen = cipher.doFinal(output, len)
        return Base64.encodeToString(output.copyOf(len + finalLen), Base64.NO_WRAP)
    }

    companion object {
        private const val TAG = "AUFEConnection"
        const val SERVER_URL = "https://vpn2.aufe.edu.cn"
        const val UAAP_LOGIN_URL =
            "http://uaap-aufe-edu-cn.vpn2.aufe.edu.cn:8118/cas/login?service=http%3A%2F%2Fjwcxk2.aufe.edu.cn%2Fj_spring_cas_security_check"
        const val TIMEOUT = 60_000L
    }
}
